Sources/SecureXPC/Server/XPCClientRequirement.swift

```diff
@@ -8,88 +8,364 @@
+            return false
+        }
+        
+        // Transform the requirements into Data form so that they can be compared
+        let requirementTransform = { (requirement: SecRequirement) -> Data? in
+            var data: CFData?
+            if SecRequirementCopyData(requirement, SecCSFlags(), &data) == errSecSuccess, let data = data as Data? {
+                return data
+            } else {
+                return nil
+            }
+        }
+        
+        // Turn into sets so they can be compared without taking into account the order of requirements
+        let requirementsData = Set<Data>(self.requirements.compactMap(requirementTransform))
+        let otherRequirementsData = Set<Data>(acceptor.requirements.compactMap(requirementTransform))
+        
+        return requirementsData == otherRequirementsData
+    }
+}
+
+/// Accepts messages that originates from a directory containing this server.
+fileprivate struct ParentBundleMessageAcceptor: MessageAcceptor {
+    /// Requests must come from a process that either has the same URL as this or is a subdirectory of it.
+    private let parentBundleURL: URL
+    
+    init(parentBundleURL: URL) {
+        self.parentBundleURL = parentBundleURL
+    }
+    
+    func acceptMessage(connection: xpc_connection_t, message: xpc_object_t) -> Bool {
+        do {
+            try expandSandboxIfNecessary(message: message)
+        } catch {
+            return false
+        }
+        
+        guard let clientCode = SecCodeCreateWithXPCConnection(connection, andMessage: message),
+              let clientPath = SecCodeCopyPath(clientCode) else {
+            return false
+        }
+        
+        // If this is true there's no possibility of the client being equal to or a subdirectory of the parent bundle.
+        // And importantly, this prevents indexing out of bounds when iterate through to check equality/containment.
+        if clientPath.pathComponents.count < parentBundleURL.pathComponents.count {
+            return false
+        }
+        
+        // Validate the each component in the client path is present in the parent bundle path in the same order
+        for i in 0..<parentBundleURL.pathComponents.count {
+            if parentBundleURL.pathComponents[i] != clientPath.pathComponents[i] {
+                return false
+            }
+        }
+        
+        return true
+    }
+    
+    private func SecCodeCopyPath(_ code: SecCode) -> URL? {
+        var staticCode: SecStaticCode?
+        guard SecCodeCopyStaticCode(code, SecCSFlags(), &staticCode) == errSecSuccess,
+              let staticCode = staticCode else {
+            return nil
+        }
+        
+        var path: CFURL?
+        guard Security.SecCodeCopyPath(staticCode, SecCSFlags(), &path) == errSecSuccess,
+              let path = (path as URL?)?.standardized else {
+            return nil
+        }
+        
+        return path
```

Comment:
This `path` shadowing is tricky, but not actually necessary here

```suggestion
        guard Security.SecCodeCopyPath(staticCode, SecCSFlags(), &path) == errSecSuccess else {
            return nil
        }
        
        return (path as URL?)?.standardized
```